SAP2000_Adapter/SAP2000Adapter.cs

```diff
@@ -56,39 +57,83 @@ public SAP2000Adapter(string filePath = "", bool active = false)
-                    openCommand.FileName = filePath;
 
-                if (System.Diagnostics.Process.GetProcessesByName("SAP2000").Length > 0)
+                if (System.Diagnostics.Process.GetProcessesByName("SAP2000").Length > 1)
                 {
-                    object runningInstance = System.Runtime.InteropServices.Marshal.GetActiveObject("CSI.SAP2000.API.SAPObject");
+                    BH.Engine.Reflection.Compute.RecordError("More than one SAP2000 instance is open. BHoM will attach to the most recently updated process, " +
+                        "but you should only work with one SAP2000 instance at a time with BHoM.");
+                }
+
+                if (File.Exists(filePath))
+                    openCommand.FileName = filePath;
 
+                // Try to attach to an existing instance of SAP2000
+                try
+                {
+                    object runningInstance = (cOAPI) System.Runtime.InteropServices.Marshal.GetActiveObject("CSI.SAP2000.API.SapObject");
                     m_app = (cOAPI)runningInstance;
+                    // Attach to current model
                     m_model = m_app.SapModel;
+                    // Continue if filepath provided
                     if (openCommand.FileName != null)
-                        RunCommand(openCommand);
+                    {
+                        FileInfo file = new FileInfo(filePath);
+                        string extension = file.Extension;
+                        if (String.Equals(extension, ".sdb"))
+                        {
+                            String openModelFileName = m_model.GetModelFilename();
+                            // Filepath matches open file
+                            if (String.Equals(openModelFileName, filePath) == false)
+                            {
+                                // Attempt to save current file, close, and open SAP2000 file
+                                m_model.File.Save();
+                                m_model.File.OpenFile(filePath);
+                                BH.Engine.Reflection.Compute.RecordWarning("A SAP2000 Model was already open, and it has been saved if it was a named file. " +
+                                    "The model at the filePath has been opened instead.");
+                            }
+                        }
+                        else
+                        {
+                            BH.Engine.Reflection.Compute.RecordError("Invalid extension. SAP2000 model filename must end with .sdb. " +
+                                "BHoM is attached to the current SAP2000 instance, but a valid SAP2000 model path has not been provided.");
+                        }
+                    }
                     else
-                        RunCommand(new NewModel());
+                    {
+                        BH.Engine.Reflection.Compute.RecordWarning("File path is either not provided or invalid. " +
```

Comment:
if SAP is open, but no model is loaded; no filePath is provided, adapter attaches but won't be able to push/pull since there's no active model - suggest adding:
```
if (m_model = null)
    RunCommand(new NewModel());